orcas/l1l2cassandra.go

```diff
@@ -0,0 +1,411 @@
+	l2  handlers.Handler
+	res protocol.Responder
+}
+
+func L1L2Cassandra(l1, l2 handlers.Handler, res protocol.Responder) orcas.Orca {
+	return &L1L2CassandraOrca{
+		l1:  l1,
+		l2:  l2,
+		res: res,
+	}
+}
+
+// This func aims at providing an async method to fill L2 after L1 set is ack'ed
+func (l *L1L2CassandraOrca) SetL2Only(req common.SetRequest) error {
+	// Set to L2
+	metrics.IncCounter(orcas.MetricCmdSetL2)
+	start := timer.Now()
+	errL2 := l.l2.Set(req)
+
+	metrics.ObserveHist(orcas.HistSetL2, timer.Since(start))
+
+	// Return an error if data is not stored in L2
+	if errL2 != nil {
+		metrics.IncCounter(orcas.MetricCmdSetErrorsL2)
+		return ErrL2OnlySetFailed
+	}
+
+	metrics.IncCounter(orcas.MetricCmdSetSuccessL2)
+	// return code is ignored by goroutine caller (Set func)
+	return errL2
+}
+
+func (l *L1L2CassandraOrca) Set(req common.SetRequest) error {
+	//log.Println("set", string(req.Key))
+
+	// Set to L1 first
+	metrics.IncCounter(orcas.MetricCmdSetL1)
+	start := timer.Now()
+	errL1 := l.l1.Set(req)
+
+	metrics.ObserveHist(orcas.HistSetL1, timer.Since(start))
+
+	if errL1 != nil {
+		metrics.IncCounter(orcas.MetricCmdSetErrorsL1)
+	} else {
+		// Successful write to L1 is a completed write !
+		metrics.IncCounter(orcas.MetricCmdSetSuccessL1)
+		metrics.IncCounter(orcas.MetricCmdSetSuccess)
+		// Set L2 asynchronously. As L1 is OK, we don't check L2 success.
+		go l.SetL2Only(req)
+		return l.res.Set(req.Opaque, req.Quiet)
+	}
+
+	// If L1 Set failed, fallback to L2 Set
+	metrics.IncCounter(orcas.MetricCmdSetL2)
+	start = timer.Now()
+	errL2 := l.l2.Set(req)
+
+	metrics.ObserveHist(orcas.HistSetL2, timer.Since(start))
+
+	// Return an error if data is not stored in L2 either
+	if errL2 != nil {
+		metrics.IncCounter(orcas.MetricCmdSetErrorsL2)
+		metrics.IncCounter(orcas.MetricCmdSetErrors)
+		return ErrL1L2SetFailed
+	}
+
+	metrics.IncCounter(orcas.MetricCmdSetSuccessL2)
+	metrics.IncCounter(orcas.MetricCmdSetSuccess)
+	return l.res.Set(req.Opaque, req.Quiet)
+}
+
+func (l *L1L2CassandraOrca) Add(req common.SetRequest) error {
+	// Add is not yet implemented.
+	log.Println("[WARN] Add command not supported by L1L2 NotConsistent orchestrator")
+	return common.ErrUnknownCmd
+}
+
+func (l *L1L2CassandraOrca) Replace(req common.SetRequest) error {
+	// Replace is not yet implemented.
+	log.Println("[WARN] Replace command not supported by L1L2 NotConsistent orchestrator")
+	return common.ErrUnknownCmd
+}
+
+func (l *L1L2CassandraOrca) Append(req common.SetRequest) error {
+	// Append is not yet implemented.
+	log.Println("[WARN] Append command not supported by L1L2 NotConsistent orchestrator")
+	return common.ErrUnknownCmd
+}
+
+func (l *L1L2CassandraOrca) Prepend(req common.SetRequest) error {
+	// Prepend is not yet implemented.
+	log.Println("[WARN] Prepend command not supported by L1L2 NotConsistent orchestrator")
+	return common.ErrUnknownCmd
+}
+
+func (l *L1L2CassandraOrca) Delete(req common.DeleteRequest) error {
```

Comment:
I think you can split the function, and use L1delete and L2delete function.
It will help readability